watcher/portfs.cpp

```diff
@@ -0,0 +1,256 @@
+struct PortFSWatcher : public Watcher {
+  FileDescriptor port_fd;
+  Pipe terminatePipe_;
+
+  /* map of file name to watchman_port_file */
+  watchman::Synchronized<
+      std::unordered_map<w_string, std::unique_ptr<watchman_port_file>>>
+      port_files;
+
+  port_event_t portevents[WATCHMAN_BATCH_LIMIT];
+
+  explicit PortFSWatcher(w_root_t* root);
+
+  std::unique_ptr<watchman_dir_handle> startWatchDir(
+      const std::shared_ptr<w_root_t>& root,
+      struct watchman_dir* dir,
+      struct timeval now,
+      const char* path) override;
+
+  bool startWatchFile(struct watchman_file* file) override;
+
+  bool consumeNotify(
+      const std::shared_ptr<w_root_t>& root,
+      PendingCollection::LockedPtr& coll) override;
+
+  bool waitNotify(int timeoutms) override;
+  void signalThreads() override;
+  bool do_watch(const w_string& name, const watchman::FileInformation& finfo);
+};
+
+static const struct flag_map pflags[] = {
+    {FILE_ACCESS, "FILE_ACCESS"},
+    {FILE_MODIFIED, "FILE_MODIFIED"},
+    {FILE_ATTRIB, "FILE_ATTRIB"},
+    {FILE_DELETE, "FILE_DELETE"},
+    {FILE_RENAME_TO, "FILE_RENAME_TO"},
+    {FILE_RENAME_FROM, "FILE_RENAME_FROM"},
+    {UNMOUNTED, "UNMOUNTED"},
+    {MOUNTEDOVER, "MOUNTEDOVER"},
+    {0, nullptr},
+};
+
+static std::unique_ptr<watchman_port_file> make_port_file(
+    const w_string& name,
+    const watchman::FileInformation& finfo) {
+  auto f = watchman::make_unique<watchman_port_file>();
+
+  f->name = name;
+  f->port_file.fo_name = (char*)name.c_str();
+  f->port_file.fo_atime = finfo.atime;
+  f->port_file.fo_mtime = finfo.mtime;
+  f->port_file.fo_ctime = finfo.ctime;
+
+  return f;
+}
+
+PortFSWatcher::PortFSWatcher(w_root_t* root)
+    : Watcher("portfs", 0), port_fd(port_create(), "port_create()") {
+  auto wlock = port_files.wlock();
+  wlock->reserve(root->config.getInt(CFG_HINT_NUM_DIRS, HINT_NUM_DIRS));
+  port_fd.setCloExec();
+}
+
+bool PortFSWatcher::do_watch(
+    const w_string& name,
+    const watchman::FileInformation& finfo) {
+  auto wlock = port_files.wlock();
+  if (wlock->find(name) != wlock->end()) {
+    // Already watching it
+    return true;
+  }
+
+  auto f = make_port_file(name, finfo);
+  auto rawFile = f.get();
+  wlock->emplace(name, std::move(f));
+
+  w_log(W_LOG_DBG, "watching %s\n", name.c_str());
+  errno = 0;
+  if (port_associate(
+          port_fd.fd(),
+          PORT_SOURCE_FILE,
+          (uintptr_t)&rawFile->port_file,
+          WATCHMAN_PORT_EVENTS,
+          (void*)rawFile)) {
+    int err = errno;
+    w_log(
+        W_LOG_ERR,
+        "port_associate %s %s\n",
+        rawFile->port_file.fo_name,
+        strerror(errno));
+    wlock->erase(name);
+    throw std::system_error(err, std::generic_category(), "port_associate");
+    return false;
+  }
+
+  return true;
+}
+
+bool PortFSWatcher::startWatchFile(struct watchman_file* file) {
+  w_string name;
+  bool success = false;
+
+  name = w_dir_path_cat_str(file->parent, file->getName());
```

Comment:
re: your comment about the delref, I'd make this line: `auto name = ...`
and then remove the explicit delref later on; `w_string` (the return type from that function) will correct destruct and free things as appropriate.  Likewise, if there are other `w_string_delref` calls in this file, you can and should make a similar change